core/src/main/scala/org/http4s/EntityDecoder.scala

```diff
@@ -99,8 +99,12 @@ object EntityDecoder extends EntityDecoderInstances {
 
   private class OrDec[T](a: EntityDecoder[T], b: EntityDecoder[T]) extends EntityDecoder[T] {
     override def decode(msg: Message): DecodeResult[T] = {
-      if (a.matchesMediaType(msg)) a.decode(msg)
-      else b.decode(msg)
+      msg.headers.get(`Content-Type`).map { contentType =>
+        if (a.matchesMediaType(contentType.mediaType)) a.decode(msg)
+        else b.decode(msg)
+      }.getOrElse {
+        a.decode(msg).orElse(b.decode(msg))
```

Comment:
I think this could be an issue: because the body is an effectful process, we shouldn't attempt to decode it twice because the first attempt will pull some bytes which cannot be put back. Therefor, if in the `getOrElse` the call `a.decode(msg)` fails, I think `b` will get a partial message body.